File: go/test/email_inbox_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/temporary-email-api2-sdk"
	"github.com/voxgig-sdk/temporary-email-api2-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestEmailInboxEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.EmailInbox(nil)
		if ent == nil {
			t.Fatal("expected non-nil EmailInboxEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := email_inboxBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "email_inbox." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TEMPORARYEMAILAPI__TEST_EMAIL_INBOX_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		emailInboxRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.email_inbox", setup.data)))
		var emailInboxRef01Data map[string]any
		if len(emailInboxRef01DataRaw) > 0 {
			emailInboxRef01Data = core.ToMapAny(emailInboxRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = emailInboxRef01Data

		// LOAD
		emailInboxRef01Ent := client.EmailInbox(nil)
		emailInboxRef01MatchDt0 := map[string]any{}
		emailInboxRef01DataDt0Loaded, err := emailInboxRef01Ent.Load(emailInboxRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if emailInboxRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func email_inboxBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "email_inbox", "EmailInboxTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read email_inbox test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse email_inbox test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"email_inbox01", "email_inbox02", "email_inbox03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TEMPORARYEMAILAPI__TEST_EMAIL_INBOX_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TEMPORARYEMAILAPI__TEST_EMAIL_INBOX_ENTID": idmap,
		"TEMPORARYEMAILAPI__TEST_LIVE":      "FALSE",
		"TEMPORARYEMAILAPI__TEST_EXPLAIN":   "FALSE",
		"TEMPORARYEMAILAPI__APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["TEMPORARYEMAILAPI__TEST_EMAIL_INBOX_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TEMPORARYEMAILAPI__TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["TEMPORARYEMAILAPI__APIKEY"],
			},
			extra,
		})
		client = sdk.NewTemporaryEmailApi2SDK(core.ToMapAny(mergedOpts))
	}

	live := env["TEMPORARYEMAILAPI__TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TEMPORARYEMAILAPI__TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/EmailInboxEntityTest.php
<?php
declare(strict_types=1);

// EmailInbox entity test

require_once __DIR__ . '/../temporaryemailapi2_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EmailInboxEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TemporaryEmailApi2SDK::test(null, null);
        $ent = $testsdk->EmailInbox(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = email_inbox_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "email_inbox." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TEMPORARY_EMAIL_API2_TEST_EMAIL_INBOX_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $email_inbox_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.email_inbox")));
        $email_inbox_ref01_data = null;
        if (count($email_inbox_ref01_data_raw) > 0) {
            $email_inbox_ref01_data = Helpers::to_map($email_inbox_ref01_data_raw[0][1]);
        }

        // LOAD
        $email_inbox_ref01_ent = $client->EmailInbox(null);
        $email_inbox_ref01_match_dt0 = [
            "id" => $email_inbox_ref01_data["id"],
        ];
        $email_inbox_ref01_data_dt0_loaded = $email_inbox_ref01_ent->load($email_inbox_ref01_match_dt0, null);
        $email_inbox_ref01_data_dt0_load_result = Helpers::to_map(is_object($email_inbox_ref01_data_dt0_loaded) && method_exists($email_inbox_ref01_data_dt0_loaded, 'data_get') ? $email_inbox_ref01_data_dt0_loaded->data_get() : $email_inbox_ref01_data_dt0_loaded);
        $this->assertNotNull($email_inbox_ref01_data_dt0_load_result);
        $this->assertEquals($email_inbox_ref01_data_dt0_load_result["id"], $email_inbox_ref01_data["id"]);

    }
}

function email_inbox_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/email_inbox/EmailInboxTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TemporaryEmailApi2SDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["email_inbox01", "email_inbox02", "email_inbox03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TEMPORARY_EMAIL_API2_TEST_EMAIL_INBOX_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TEMPORARY_EMAIL_API2_TEST_EMAIL_INBOX_ENTID" => $idmap,
        "TEMPORARY_EMAIL_API2_TEST_LIVE" => "FALSE",
        "TEMPORARY_EMAIL_API2_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TEMPORARY_EMAIL_API2_TEST_EMAIL_INBOX_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TEMPORARY_EMAIL_API2_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new TemporaryEmailApi2SDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TEMPORARY_EMAIL_API2_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TEMPORARY_EMAIL_API2_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
